### DISCUSSION TITLE
음파 통신 속도 2배를 위한 사전 스터디

### DISCUSSION THREAD

**Discussion by dlwlstks96** [euphony-io/euphony#52] [General]
![마스터코스_미션](https://user-images.githubusercontent.com/72801465/133914336-9d361831-dd48-4b37-a1b0-afc10bc0df01.JPG)

멘토님께서 보내주셨던 euphony 음파 통신 속도 2배 향상 관련 내용입니다!
팀 구분 없이 모두가 보시면 euphony 접근에 도움이 되실 것 같아 디스커션 남깁니다!
1. FFT와 관련된 참고 영상 - [FFT(푸리에 변환)](https://www.youtube.com/watch?v=spUNpyF58BY)
2. 위 이미지에 나와 있는 BASE 인터페이스의 경로
`euphony - src - main - cpp - core - Base.h`

저도 차근차근 공부 중이라 해당 내용에 관련된 좋은 자료가 있다면 공유해주시면 감사하겠습니다:)

**Comment by YoungSeokHong** [comment]
와 푸리에 변환이 너무 큰 산이였는데 영상보니까 한 번에 이해가 되네요! 진짜 잘 정리되어 있는 영상인 것 같습니다. 감사합니다!

**Comment by YoungSeokHong** [comment]
### 1. FFT사이즈와 속도의 관계

FFT 사이즈와 해상도의 관계에 대해서 공부하던 중 괜찮은 강의가 있어서 공유합니다!
 -> [FFT 사이즈와 해상도 변화](https://www.youtube.com/watch?v=72oR-LpbHyo)

공부한 내용을 바탕으로 이해해보자면
1. Base32를 구현하고 FFT사이즈를 1024로 설정
2. FFT사이즈의 증가 -> 주파수 분해능이 작아짐 = 음파를 더욱 세밀하게 분석할 수 있어짐
3. 음파를 더욱 세밀하게 분석할 수 있으므로 그만큼 많은 데이터를 한 번에 전송 가능 -> 속도의 증가

이런 식으로 이해했는데 이게 맞을까요??

### 2. RX 코드에 대한 정리 

그리고 코드 RX코드 분석은 이전에 이슈에 올라왔던 내용을 참고해서 직접 적어볼까 하다가 이전에 올라왔던 이슈가 잘 정리되어 있는 것 같아서 일단 공유합니다!
-> [코드 설명 - 클래스를 중심으로](https://github.com/euphony-io/euphony/issues/18#issuecomment-897895025)

우선은 결국에 RX에서 가장 중요한 클래스를 꼽자면 KissFFT와 EuRxManager인 것 같습니다.
EuRxManager는 대부분의 멘티님들이 사용해보셨으면 아시겠지만 외부에서 listen()함수를 불러오면 스레드를 하나 생성해 RxRunner()를 작동시키고 finish()함수로 스레드를 종료시킵니다.  RxRunner는 EuFreqObject를 상속받은 class로 스레드가 종료될 때까지 FFT등의 정보 수신에 필요한 작업을 진행합니다.
KissFFT는 C++ 모듈을 사용하는 class로 고속 푸리에 변환을 작동하도록 모듈을 동작시키는 class입니다. euphony/cpp/native-legacy.cpp 에서 이 모듈에 관련된 코드를 확인하실 수 있습니다.

> 제가 잘 못 이해한 내용이 있으면 댓글로 남겨주세요. 피드백은 언제나 환영합니다!
> EUPI에 대해서 공부하고 있는데 EUPI는 정확히 어떤 역할을 하는지 잘 모르겠네요 ㅠㅠ KEY_DOWN, KEY_UP, KEY_PRESSED으로 이루어진 enum들도 어떤 상황에 사용되는지 조금더 공부해봐야될 것 같습니다. 혹시 아시는 분 있으시면 공유 부탁드립니다!

**Comment by YoungSeokHong** [comment]
### base32를 만들기 위한 준비

결국 음파 통신의 속도를 늘리기 위해선 base32를 만들어야 되는데, 저에게도 그렇듯이 base.h가 많이 낯설으실 것같습니다ㅠㅠ base.h를 공부하실 때 이걸 참고하시면 도움이 될 것 같아서 가져와봤습니다.

base32를 만들기 위해선 base16, base2를 참고하면 좋을 것 같은데 cpp\core\base16.h와 cpp\core\base2.h를 테스트하기 위한 base2Test.cpp와 base16Test.cpp가 아시다시피 cpp\tests 디렉토리에 위치합니다!
![image](https://user-images.githubusercontent.com/80624082/134507426-67f7d3ac-6386-4fb5-acbd-206f72831ccb.png)
![image](https://user-images.githubusercontent.com/80624082/134507493-d50b8051-8bf9-4d1f-99ed-895882045ca7.png)

저번 챌린지코스 때 멘토님께서 test 케이스를 잘 뜯어보면 코드의 전반적인 이해가 가능하다고 하신 것처럼 base2, base16의 테스트 케이스를 분석해두면 base를 이해하는데 조금은 도움이 될 것같습니다! 저도 아직은 분석해보진 않았지만, 시간이 나는대로 분석해서 열심히 코멘트하겠습니다! 관심이 있으신 멘티님들은 같이 테스트 코드를 뜯어보면서 이곳에서 의견 나누면 좋을 것 같습니다.

**Reply by designe** [reply]
Base32 관련하여 디스커션 분기해도 좋을 것 같네요!

**Reply by YoungSeokHong** [reply]
#66 으로 새로 만들었습니다!
